sstables/sstable_set.cc

```diff
@@ -765,6 +768,403 @@ sstable_set_impl::create_single_key_sstable_reader(
+    }
+
+    // Relevant only in forwarding mode (in non-forwarding mode _halted_readers is always empty).
+    // Remove all halted readers (i.e. readers which returned end-of-stream) that we know for sure
+    // won't return any more fragments. We use the sstables' max_position() metadata to make the decision.
+    // Precondition: every halted reader returned all fragments with positions <= `up_to`.
+    // See also the comment for `_halted_readers`.
+    void collect_exhausted_readers(position_in_partition_view up_to) {
+        while (!_halted_readers.empty()
+                && _cmp(up_to, _halted_readers.front()->max_position) >= 0) {
+            boost::range::pop_heap(_halted_readers, _halted_reader_cmp);
+            _all_readers.erase(_halted_readers.back());
+            _halted_readers.pop_back();
+        }
+    }
+
+    // Retrieve the next fragment from the reader pointed to by `it`.
+    // The function assumes that we're not in galloping mode, `it` is in `_unpeeked_readers`,
+    // and all fragments previously returned from the reader have already been returned by operator().
+    //
+    // The peeked reader is pushed onto the _peeked_readers heap.
+    future<> peek_reader(reader_iterator it, db::timeout_clock::time_point timeout) {
+        return it->reader.peek(timeout).then([this, timeout, it] (mutation_fragment* mf) {
+            if (!mf) {
+                // The reader returned end-of-stream before returning end-of-partition
+                // (otherwise we would have removed it in a previous peek). This means we are in forwarding mode.
+                // We don't know if the reader won't start returning fragments again, so we save it
+                // for the moment. If it later turns out that we've already walked past the reader's
+                // max_position(), we will remove it (see `collect_exhausted_readers`). Otherwise,
+                // if we get fast-forwarded, we will fast-forward it and bring it back.
+                _halted_readers.push_back(it);
+                boost::range::push_heap(_halted_readers, _halted_reader_cmp);
+                return make_ready_future<>();
+            }
+
+            if (mf->is_partition_start()) {
+                // We assume there are no partition tombstones.
+                // This should have been checked before opening the reader.
+                if (mf->as_partition_start().partition_tombstone()) {
+                    on_internal_error(setlogger, format(
+                            "clustering_order_reader_merger: partition tombstone encountered for partition {}."
+                            " This sstable reader cannot be used for sstables that have partition tombstones"
+                            " or it would give incorrect results. The no-tombstone condition should have been checked"
+                            " before opening the reader. Please report a bug.", mf->as_partition_start().key()));
```

Comment:
"The no-tombstone condition should have been checked before opening the reader. Please report a bug." should be removed. We could append this to all error messages :D.

Reply:
Well, only for errors that are due to an internal bug (so these errors should never happen, like this one), not "user error" messages (which will happen in normal circumstances).

Reply:
Still, we don't append such text to no other `assert()`/`on_internal_error()`, etc. Users might start believing they only need to report a bug if the message asks them to.